build_resolvers/lib/src/resolver.dart

```diff
@@ -60,11 +61,20 @@ class PerActionResolver implements ReleasableResolver {
       libraries.firstWhere((l) => l.name == libraryName, orElse: () => null);
 
   @override
-  Future<bool> isLibrary(AssetId assetId) => _delegate.isLibrary(assetId);
+  Future<bool> isLibrary(AssetId assetId) =>
+      _resolveIfNecesssary(assetId).then(_delegate.isLibrary);
 
   @override
   Future<LibraryElement> libraryFor(AssetId assetId) =>
-      _delegate.libraryFor(assetId);
+      _resolveIfNecesssary(assetId).then(_delegate.libraryFor);
+
+  Future<AssetId> _resolveIfNecesssary(AssetId id) async {
+    if (!_delegate._uriResolver.seenAssets.contains(id)) {
```

Comment:
`seenAssets` is not specific to a `buildStep` so this could cause use to expose information without properly calling `canRead` for this step. I think we need to somehow have a per-build step understanding of what we have already crawled...

Reply:
Ya, I didn't realize this wasn't per build step - not guarding it behind this check would also mean a regression for _all_ calls to `libraryFor` so we can't just omit this either.

If we do track this per build step we can potentially make `performResolve` more efficient too, so it doesn't re-crawl things its already crawled in the same step.